controllers/kvmcluster_controller.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/annotations"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/giantswarm/cluster-api-provider-kvm/api/v1alpha4"
+	"github.com/giantswarm/cluster-api-provider-kvm/pkg/kvm/metrics"
+	"github.com/giantswarm/cluster-api-provider-kvm/pkg/kvm/scope"
+	"github.com/giantswarm/cluster-api-provider-kvm/pkg/kvm/services/namespace"
+	"github.com/pkg/errors"
+)
+
+const (
+	KVMClusterController = "kvmcluster"
+)
+
+// KVMClusterReconciler reconciles a KVMCluster object
+type KVMClusterReconciler struct {
+	client.Client
+	Recorder         record.EventRecorder
+	Scheme           *runtime.Scheme
+	WatchFilterValue string
+}
+
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io.infrastructure.cluster.x-k8s.io,resources=kvmclusters/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the KVMCluster object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *KVMClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reconcileError error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	defer metrics.CaptureLastReconciled(KVMClusterController)
```

Comment:
It seems like this will always update the last-reconciled metric even when there's an error. Do we consider it to be a reconciliation if it fails?

Reply:
Good point, updated.